smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,462 @@
+                        self.logger.debug("Budget exhausted by adaptive capping; "
+                                          "Interrupting current challenger and moving on to the next one")
+                        break
+
+                    except BudgetExhaustedException:
+                        # Returning the final incumbent selected so far because we ran out of optimization budget
+                        self.logger.debug("Budget exhausted; "
+                                          "Interrupting optimization run and returning current incumbent")
+                        return incumbent, None
+
+                    self._num_run += 1
+
+                if not first_run:
+                    first_run = True
+                    self._chall_indx += 1
+
+            # selecting the top 'k' challengers for the next iteration
+            next_n_chal = int(np.round(len(curr_challengers) / self.eta))
+            if next_n_chal >= 1:
+                curr_challengers = self._top_k(curr_challengers, run_history, k=next_n_chal)
+            else:
+                # returning at least 1 configuration for next iteration since the incumbent has to run on all instances
+                self.logger.debug('Ran out of configurations! Returning the best from what was seen so far.')
+                curr_challengers = self._top_k(curr_challengers, run_history, k=1)
+
+            # increase budget for next iteration
+            prev_budget = budget
+            budget = budget * self.eta
+
+        # select best challenger from the SH run
+        best_challenger = curr_challengers[0]
+        self.logger.debug("Best challenger from successive halving run - %s" % (str(best_challenger)))
+
+        # compare challenger with current incumbent
+        new_incumbent, inc_perf = self._compare_configs(incumbent, best_challenger,
+                                                        run_history, aggregate_func, log_traj)
+
+        return new_incumbent, inc_perf
+
+    def _top_k(self, configs: typing.List[Configuration], run_history: RunHistory, k: int):
+        """
+        selects the top 'k' configurations from the given list based on their performance in this budget
+
+        Parameters
+        ----------
+        configs: typing.List[Configuration]
+            list of configurations to filter from
+        run_history: RunHistory
+            stores all runs we ran so far
+        k: int
+            number of configurations to select
+
+        Returns
+        -------
+        challengers: typing.List[Configuration]
+            top challenger configurations, sorted in increasing costs
+        """
+        # extracting costs for each given configuration
+        config_costs = {}
+        for c in configs:
+            runs = run_history.get_runs_for_config(c)
+            perf = run_history.aggregate_func(c, run_history, runs)
+            config_costs[c] = perf
+
+        configs_sorted = sorted(config_costs, key=config_costs.get)
```

Comment:
what would happen if we consider a configuration for a second time? I worry that we will compare configurations based on a different kinds of sets. I think we have to limit the considered `runs` performed in the current SH run

Reply:
In the current implementation, we remove configurations that have already been run at least once before starting the successive halving iterations. So we are not considering a configuration for the 2nd time in SH. 

Configurations being compared at this stage are only those that have run this given set of instance-seed pairs in the current budget. So `runs` would be the same for all configs. (I could/should take it out of the loop)

However, if cutoff is used as a budget, configurations will repeat i.e., on larger budgets, we will run a configuration on the same instance-seed pair again. But `RunHistory` overwrites the cost for that key in its dictionary, so the configurations stay comparable.